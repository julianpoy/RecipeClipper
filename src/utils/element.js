export const getClassNamesMatching = (classNamePartial) => {
  const classRegExp = new RegExp(`class="((\\w|\\s|-)*${classNamePartial}(\\w|\\s|-)*)"`, 'gi');
  const matches = document.body.innerHTML.matchAll(classRegExp);

  return Array.from(new Set(Array.from(matches, (match) => match[1])));
};

export const softMatchElementsByClass = (classNamePartial) => {
  const classNames = getClassNamesMatching(classNamePartial);

  return classNames
    .map((className) => Array.from(document.getElementsByClassName(className)))
    .flat();
};

export const applyLIBlockStyling = (element) => {
  Array.from(element.querySelectorAll('li')).forEach((li) => {
    li.style.display = 'block';
  });

  return element;
};

export const grabLongestMatchByClasses = (preferredClassNames, fuzzyClassNames) => {
  const exactMatches = preferredClassNames
    .map((className) => Array.from(document.getElementsByClassName(className)))
    .flat();
  const fuzzyMatches = fuzzyClassNames
    .map((className) => softMatchElementsByClass(className))
    .flat();

  return (exactMatches.length > 0 ? exactMatches : fuzzyMatches)
    .map((element) => applyLIBlockStyling(element))
    .map((element) => element.innerText.trim())
    .reduce((max, match) => (match.length > max.length ? match : max), '');
};

export const isImg = (element) => element.tagName.toLowerCase().trim() === 'img';
export const isPicture = (element) => element.tagName.toLowerCase().trim() === 'picture';

export const getImgElementsWithin = (element) => {
  const matchedImgElements = [];
  if (isImg(element)) matchedImgElements.push(element);
  matchedImgElements.push(...element.querySelectorAll('img'));
  return matchedImgElements;
};

export const getAttrIfExists = (el, attrName) => {
  if (el.attributes[attrName]) return el.attributes[attrName].value;
  return '';
};

export const getSrcFromImage = (img) => {
  if (!img) return '';

  const closestSrc = getAttrIfExists(img, 'data-src') || getAttrIfExists(img, 'data-lazy-src') || img.currentSrc || img.src;
  return closestSrc || '';
};

export const isValidImage = (element) => isImg(element)
    && !!getSrcFromImage(element)
    && element.complete // Filter images that haven't completely loaded
    && element.naturalWidth > 0 // Filter images that haven't loaded correctly
    && element.naturalHeight > 0;

export const getImageDimensions = (element) => {
  const parent = element.parentNode;
  const isParentPicture = parent && isPicture(parent);
  const offsetHeight = isParentPicture
    ? Math.max(element.offsetHeight, parent.offsetHeight) : element.offsetHeight;
  const offsetWidth = isParentPicture
    ? Math.max(element.offsetWidth, parent.offsetWidth) : element.offsetWidth;

  return {
    offsetHeight,
    offsetWidth,
  };
};

export const grabLargestImage = () => {
  const matches = document.querySelectorAll('img');

  return [...matches]
    .filter((element) => isValidImage(element))
    .reduce((max, element) => {
      const { offsetWidth, offsetHeight } = getImageDimensions(element);

      // Do not use images smaller than 200x200
      if (offsetWidth < 200 && offsetHeight < 200) return max;

      const elTotalPx = offsetHeight * offsetWidth;
      const maxTotalPx = max ? (max.offsetHeight * max.offsetWidth) : 0;

      return elTotalPx > maxTotalPx ? element : max;
    }, null);
};

export const closestToRegExp = (regExp) => {
  const match = document.body.innerText.match(regExp);
  if (!match) return '';
  return match[0];
};

export const grabRecipeTitleFromDocumentTitle = () => document.title.split(/ - |\|/)[0].trim();

export const grabSourceFromDocumentTitle = () => (document.title.split(/ - |\|/)[1] || '').trim();
