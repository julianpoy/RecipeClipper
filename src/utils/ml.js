// Self import for mock
import * as self from './ml';
import { getInnerText } from './innerText';
import {
  applyLIBlockStyling,
} from './element';
import {
  ingredientSectionHeader,
  instructionSectionHeader,
} from '../constants/regex';

export const getDocumentContent = (config) => {
  const { body } = config.window.document;
  applyLIBlockStyling(body);

  return getInnerText(body).split('\n').map((line) => line.trim()).filter((line) => line);
};

export const findPotentialSetsByHeader = (config, headerRegexp) => {
  const content = self.getDocumentContent(config);

  return content.filter((line) => line.match(headerRegexp))
    .map((line) => content.slice(content.indexOf(line) + 1));
};

export const loadModel = async (config) => {
  const modelUrl = config.options.mlModelEndpoint;
  if (!modelUrl) throw new Error('You must provide window.RC_ML_MODEL_ENDPOINT or options.mlModelEndpoint to use local classification');
  return config.window.tf.loadLayersModel(modelUrl);
};

export const mlClassifyLocal = async (config, lines) => {
  const model = await self.loadModel(config);
  const useModel = await config.window.use.load();

  const predictions = [];
  for (let i = 0; i < lines.length; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    const encodedData = await useModel.embed([lines[i]]);
    const prediction = model.predict(encodedData).dataSync();
    predictions.push(prediction);
  }

  return predictions;
};

export const mlClassifyRemote = async (config, lines) => {
  const remote = config.options.mlClassifyEndpoint;
  if (!remote) throw new Error('You must provide window.RC_ML_CLASSIFY_ENDPOINT or options.mlClassifyEndpoint to use remote classification');

  const response = await config.window.fetch(remote, {
    method: 'POST',
    body: JSON.stringify({
      sentences: lines,
    }),
    headers: {
      'Content-Type': 'application/json',
    },
  });

  return response.json();
};

export const mlClassify = async (config, lines) => {
  const isTFJSAvailable = config.window.tf && config.window.tf.loadLayersModel;
  const isUSEAvailable = config.window.use && config.window.use.load;

  if (isTFJSAvailable && isUSEAvailable) return self.mlClassifyLocal(config, lines);
  return self.mlClassifyRemote(config, lines);
};

export const mlFilter = async (config, lines, type) => {
  const predictions = await self.mlClassify(config, lines);

  let lastType = -1;
  const filteredOutput = [];
  for (let i = 0; i < lines.length; i += 1) {
    const predictedType = predictions[i].indexOf(Math.max(...predictions[i])) + 1;

    // Allow one line of error to be included
    // (say, an ingredient header or a stray bit of formatting)
    // Also allow some non-ingredient text at the very start (say, a section header)
    if (predictedType === type || lastType === type || i === 0) {
      filteredOutput.push(lines[i]);
      lastType = predictedType;
    } else {
      break;
    }
  }

  // Remove last element if it is not of the desired type
  // (consider our allowed buffer above may insert one undesired item at end)
  if (filteredOutput.length > 0 && lastType !== type) {
    filteredOutput.pop();
  }

  return filteredOutput;
};

export const findFullSearch = async (config, type) => {
  const content = self.getDocumentContent(config);

  const predictions = await self.mlClassify(config, content);

  const { groups, workingGroup } = content.reduce((acc, line, idx) => {
    const predictedType = predictions[idx].indexOf(Math.max(...predictions[idx])) + 1;

    const lastType = acc.workingGroup
      && acc.workingGroup.length > 0 ? acc.workingGroup[acc.workingGroup.length - 1].type : -1;

    // Allow one line of error to be included in group
    // (say, an ingredient header or a stray bit of formatting)
    // Also allow some non-ingredient text at the very start (say, a section header)
    if (predictedType === type || lastType === type || idx === 0) {
      acc.workingGroup = acc.workingGroup || [];
      acc.workingGroup.push({
        text: line,
        type: predictedType,
      });
    } else {
      // Group is considered complete
      if (acc.workingGroup && acc.workingGroup.length > 0) acc.groups.push(acc.workingGroup);
      acc.workingGroup = null;
    }

    return acc;
  }, {
    workingGroup: null,
    groups: [],
  });

  // If leftover working group (if ingredients/instructions end at very end of page)
  if (workingGroup) groups.push(workingGroup);

  groups.forEach((group) => {
    // Remove last element if it is not of the desired type
    // (consider our allowed buffer above may insert one undesired item at end)
    if (group.length > 0 && group[group.length - 1].type !== type) {
      group.pop();
    }
  });

  return groups
    .map((group) => group.map((item) => item.text))
    .map((group) => group.join('\n'))
    .reduce((a, b) => (a.length > b.length ? a : b), '');
};

export const findByHeader = async (config, type) => {
  const headerRegexp = type === 1 ? ingredientSectionHeader : instructionSectionHeader;
  const potentialSets = self.findPotentialSetsByHeader(config, headerRegexp);

  const sets = [];
  for (let i = 0; i < potentialSets.length; i += 1) {
    const potentialSet = potentialSets[i];
    // eslint-disable-next-line no-await-in-loop
    const set = await self.mlFilter(config, potentialSet, type);
    sets.push(set);
  }

  return sets.map((set) => set.join('\n'))
    .reduce((a, b) => (a.length > b.length ? a : b), '');
};

export const find = async (config, type) => {
  if (config.options.mlDisable) return '';

  const result = await self.findByHeader(config, type) || await self.findFullSearch(config, type);

  return result;
};

// Type 1 for ingredients
// Type 2 for instructions
// Others to be implemented in future...
export const grabByMl = async (config, type) => {
  try {
    return await self.find(config, type);
  } catch (e) {
    if (config.options.ignoreMLClassifyErrors) {
      return '';
    }

    throw e;
  }
};
