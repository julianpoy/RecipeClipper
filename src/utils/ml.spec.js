import * as element from './element';
import * as ml from './ml';
import {
  grabByMl,
  findByHeader,
  findFullSearch,
  mlFilter,
  mlClassify,
  mlClassifyRemote,
  mlClassifyLocal,
  loadModel,
  findPotentialSetsByHeader,
  getDocumentContent,
} from './ml';
import {
  ingredientSectionHeader,
  instructionSectionHeader,
} from '../constants/regex';

const config = {
  window,
  options: {},
};

describe('ml', () => {
  afterEach(() => {
    jest.restoreAllMocks();
    config.options.mlDisable = undefined;
  });

  describe('grabByMl', () => {
    let findByHeaderMock;
    let findFullSearchMock;

    beforeEach(() => {
      findByHeaderMock = jest.spyOn(ml, 'findByHeader');
      findFullSearchMock = jest.spyOn(ml, 'findFullSearch');
    });

    describe('when mlDisable true', () => {
      let result;

      beforeEach(async () => {
        config.options.mlDisable = true;
        result = await grabByMl(config, 1);
      });

      it('returns empty string', () => {
        expect(result).toEqual('');
      });

      it('does not call findByHeader', () => {
        expect(ml.findByHeader).not.toHaveBeenCalled();
      });

      it('does not call findFullSearch', () => {
        expect(ml.findFullSearch).not.toHaveBeenCalled();
      });
    });

    describe('when findByHeader finds a match', () => {
      let result;
      const mockHeaderResult = 'headerExample';

      beforeEach(async () => {
        findByHeaderMock.mockResolvedValue(mockHeaderResult);

        result = await grabByMl(config, 1);
      });

      it('returns match from findByHeader', () => {
        expect(result).toEqual(mockHeaderResult);
      });

      it('calls findByHeader', () => {
        expect(ml.findByHeader).toHaveBeenCalledWith(config, 1);
      });

      it('does not call findFullSearch', () => {
        expect(ml.findFullSearch).not.toHaveBeenCalled();
      });
    });

    describe('when findByHeader does not find a match', () => {
      let result;
      const mockFullTextResult = 'fullTextExample';

      beforeEach(async () => {
        findByHeaderMock.mockReturnValue(null);
        findFullSearchMock.mockReturnValue(mockFullTextResult);

        result = await grabByMl(config, 1);
      });

      it('returns match from findFullSearch', () => {
        expect(result).toEqual(mockFullTextResult);
      });

      it('calls findByHeader', () => {
        expect(ml.findByHeader).toHaveBeenCalledWith(config, 1);
      });

      it('calls findFullSearch', () => {
        expect(ml.findFullSearch).toHaveBeenCalledWith(config, 1);
      });
    });
  });

  describe('findByHeader', () => {
    let findPotentialSetsByHeaderMock;
    const findPotentialSetsByHeaderMockReturn = [
      ['example line', 'example line 2'],
      ['example 2 line', 'example 2 line 2'],
    ];
    let mlFilterMock;
    let result;

    beforeEach(() => {
      findPotentialSetsByHeaderMock = jest.spyOn(ml, 'findPotentialSetsByHeader')
        .mockReturnValue(findPotentialSetsByHeaderMockReturn);
      mlFilterMock = jest.spyOn(ml, 'mlFilter');
    });

    describe('for ingredients', () => {
      beforeEach(async () => {
        mlFilterMock
          .mockReturnValueOnce(findPotentialSetsByHeaderMockReturn[0])
          .mockReturnValueOnce(findPotentialSetsByHeaderMockReturn[1]);
        result = await findByHeader(config, 1);
      });

      it('calls findPotentialSetsByHeader with ingredient regexp', () => {
        expect(findPotentialSetsByHeaderMock).toHaveBeenCalledWith(config, ingredientSectionHeader);
      });

      it('calls mlFilter with ingredient type', () => {
        expect(mlFilterMock.mock.calls).toEqual([
          [config, findPotentialSetsByHeaderMockReturn[0], 1],
          [config, findPotentialSetsByHeaderMockReturn[1], 1],
        ]);
      });
    });

    describe('for instructions', () => {
      beforeEach(async () => {
        mlFilterMock
          .mockReturnValueOnce(findPotentialSetsByHeaderMockReturn[0])
          .mockReturnValueOnce(findPotentialSetsByHeaderMockReturn[1]);
        result = await findByHeader(config, 2);
      });

      it('calls findPotentialSetsByHeader with instruction regexp', () => {
        expect(findPotentialSetsByHeaderMock)
          .toHaveBeenCalledWith(config, instructionSectionHeader);
      });

      it('calls mlFilter with ingredient type', () => {
        expect(mlFilterMock.mock.calls).toEqual([
          [config, findPotentialSetsByHeaderMockReturn[0], 2],
          [config, findPotentialSetsByHeaderMockReturn[1], 2],
        ]);
      });
    });

    describe('for ingredients and instructions', () => {
      beforeEach(async () => {
        mlFilterMock
          .mockReturnValueOnce(findPotentialSetsByHeaderMockReturn[0])
          .mockReturnValueOnce(findPotentialSetsByHeaderMockReturn[1]);
        result = await findByHeader(config, 2);
      });

      it('returns the longest length result as a string', () => {
        expect(result).toEqual(findPotentialSetsByHeaderMockReturn[1].join('\n'));
      });
    });

    describe('for ingredients and instructions reverse order', () => {
      beforeEach(async () => {
        mlFilterMock
          .mockReturnValueOnce(findPotentialSetsByHeaderMockReturn[1])
          .mockReturnValueOnce(findPotentialSetsByHeaderMockReturn[0]);
        findPotentialSetsByHeaderMock
          .mockReturnValue(findPotentialSetsByHeaderMockReturn.slice().reverse());
        result = await findByHeader(config, 2);
      });

      it('returns the longest length result as a string', () => {
        expect(result).toEqual(findPotentialSetsByHeaderMockReturn[1].join('\n'));
      });
    });

    describe('no content', () => {
      beforeEach(async () => {
        mlFilterMock.mockReturnValue([]);
        result = await findByHeader(config, 2);
      });

      it('returns nothing', () => {
        expect(result).toEqual('');
      });
    });
  });

  describe('findFullSearch', () => {
    let getDocumentContentMock;
    const getDocumentContentMockReturn = [
      '1 tbsp butter', // Normal ingredient
      'buy now', // Button in the middle of the group - should be forgiven
      '1 cup milk',
      '1 cup flour',

      'my life story...', // Some non-ingredient and non-instruction text
      'more life story...',
      'even more life story...',

      'Flour 50% off!', // A lone line that, say, is interpereted as an ingredient

      'Add butter to the mixing bowl', // Normal instruction
      'Add milk to the mixing bowl',
      'Add flour to the mixing bowl',
    ];
    let mlClassifyMock;
    const mlClassifyMockReturn = [
      [1, 0, 0], // Ingredient
      [0, 0, 1], // Non-ingredient
      [1, 0, 0], // Ingredient
      [1, 0, 0],

      [0, 0, 1], // Non-recipe
      [0, 0, 1],
      [0, 0, 1],

      [1, 0, 0], // Ingredient

      [0, 1, 0], // Instruction
      [0, 1, 0],
      [0, 1, 0],
    ];

    beforeEach(() => {
      getDocumentContentMock = jest.spyOn(ml, 'getDocumentContent').mockReturnValue(getDocumentContentMockReturn);
      mlClassifyMock = jest.spyOn(ml, 'mlClassify').mockReturnValue(mlClassifyMockReturn);
    });

    it('sanity/setup test - test arrays equal length', () => {
      expect(getDocumentContentMockReturn.length).toEqual(mlClassifyMockReturn.length);
    });

    it('calls getDocumentContent', async () => {
      await findFullSearch(config, 1);
      expect(getDocumentContentMock).toHaveBeenCalledWith(config);
    });

    it('calls mlClassify', async () => {
      await findFullSearch(config, 1);
      expect(mlClassifyMock).toHaveBeenCalled();
    });

    it('returns largest group of ingredients', async () => {
      const expectedIngredients = getDocumentContentMockReturn.slice(0, 4).join('\n');

      expect(await findFullSearch(config, 1)).toEqual(expectedIngredients);
    });

    it('returns largest group of instructions', async () => {
      const expectedInstructions = getDocumentContentMockReturn.slice(8, 11).join('\n');

      expect(await findFullSearch(config, 2)).toEqual(expectedInstructions);
    });
  });

  describe('mlFilter', () => {
    let mlClassifyMock;

    beforeEach(() => {
      mlClassifyMock = jest.spyOn(ml, 'mlClassify');
    });

    describe('ingredients', () => {
      const content = [
        '1 tbsp butter', // Normal ingredient
        'buy now', // Button in the middle of the group - should be forgiven
        '1 cup milk',
        '1 cup flour',

        'my life story...', // Some non-ingredient and non-instruction text
        'more life story...',
        'even more life story...',

        'Add butter to the mixing bowl', // Normal instruction
        'Add milk to the mixing bowl',
        'Add flour to the mixing bowl',

        'Flour 50% off!', // A lone line that, say, is interpereted as an ingredient
      ];
      const mlClassifyMockReturn = [
        [1, 0, 0], // Ingredient
        [0, 0, 1], // Non-ingredient
        [1, 0, 0], // Ingredient
        [1, 0, 0],

        [0, 0, 1], // Non-recipe
        [0, 0, 1],
        [0, 0, 1],

        [0, 1, 0], // Instruction
        [0, 1, 0],
        [0, 1, 0],

        [1, 0, 0], // Ingredient
      ];

      beforeEach(() => {
        mlClassifyMock.mockReturnValue(mlClassifyMockReturn);
      });

      it('returns ingredients up to non-ingredient content', async () => {
        const expectedIngredients = content.slice(0, 4);

        expect(await mlFilter(config, content, 1)).toEqual(expectedIngredients);
      });
    });

    describe('instructions', () => {
      const content = [
        'Add butter to the mixing bowl', // Normal instruction
        'Add milk to the mixing bowl',
        'Add flour to the mixing bowl',

        '1 tbsp butter', // Normal ingredient
        'buy now', // Button in the middle of the group - should be forgiven
        '1 cup milk',
        '1 cup flour',

        'my life story...', // Some non-ingredient and non-instruction text
        'more life story...',
        'even more life story...',

        'Flour 50% off!', // A lone line that, say, is interpereted as an ingredient
      ];
      const mlClassifyMockReturn = [
        [0, 1, 0], // Instruction
        [0, 1, 0],
        [0, 1, 0],

        [1, 0, 0], // Ingredient
        [0, 0, 1], // Non-ingredient
        [1, 0, 0], // Ingredient
        [1, 0, 0],

        [0, 0, 1], // Non-recipe
        [0, 0, 1],
        [0, 0, 1],

        [1, 0, 0], // Ingredient
      ];

      beforeEach(() => {
        mlClassifyMock.mockReturnValue(mlClassifyMockReturn);
      });

      it('returns instructions up to non-instruction content', async () => {
        const expectedInstructions = content.slice(0, 3);

        expect(await mlFilter(config, content, 2)).toEqual(expectedInstructions);
      });
    });

    describe('neither instructions nor ingredients', () => {
      const content = [
        'my life story...', // Some non-ingredient and non-instruction text
        'more life story...',
        'even more life story...',

        'Add butter to the mixing bowl', // Normal instruction
        'Add milk to the mixing bowl',
        'Add flour to the mixing bowl',

        '1 tbsp butter', // Normal ingredient
        'buy now', // Button in the middle of the group - should be forgiven
        '1 cup milk',
        '1 cup flour',

        'Flour 50% off!', // A lone line that, say, is interpereted as an ingredient
      ];
      const mlClassifyMockReturn = [
        [0, 0, 1], // Non-recipe
        [0, 0, 1],
        [0, 0, 1],

        [0, 1, 0], // Instruction
        [0, 1, 0],
        [0, 1, 0],

        [1, 0, 0], // Ingredient
        [0, 0, 1], // Non-ingredient
        [1, 0, 0], // Ingredient
        [1, 0, 0],

        [1, 0, 0], // Ingredient
      ];

      beforeEach(() => {
        mlClassifyMock.mockReturnValue(mlClassifyMockReturn);
      });

      it('returns nothing for ingredient filter', async () => {
        expect(await mlFilter(config, content, 1)).toEqual([]);
      });

      it('returns nothing for instruction filter', async () => {
        expect(await mlFilter(config, content, 2)).toEqual([]);
      });
    });

    describe('no content', () => {
      const content = [];
      const mlClassifyMockReturn = [];

      beforeEach(() => {
        mlClassifyMock.mockReturnValue(mlClassifyMockReturn);
      });

      it('returns nothing', async () => {
        expect(await mlFilter(config, content, 1)).toEqual([]);
      });
    });
  });

  describe('mlClassify', () => {
    const input = [
      'line 1',
      'line 2',
    ];
    let mlClassifyRemoteMock;
    const mlClassifyRemoteMockReturn = [
      [1, 0, 0],
    ];
    let mlClassifyLocalMock;
    const mlClassifyLocalMockReturn = [
      [0, 0, 1],
    ];

    beforeEach(() => {
      mlClassifyRemoteMock = jest.spyOn(ml, 'mlClassifyRemote').mockResolvedValue(mlClassifyRemoteMockReturn);
      mlClassifyLocalMock = jest.spyOn(ml, 'mlClassifyLocal').mockResolvedValue(mlClassifyLocalMockReturn);
      window.tf = undefined;
      window.use = undefined;
    });

    describe('when USE is present', () => {
      describe('when tensorflow is present', () => {
        let result;

        beforeEach(async () => {
          window.tf = {
            loadLayersModel: () => {},
          };
          window.use = {
            load: () => {},
          };

          result = await mlClassify(config, input);
        });

        it('calls local classify', () => {
          expect(mlClassifyLocalMock).toHaveBeenCalledWith(config, input);
        });

        it('does not call remote classify', () => {
          expect(mlClassifyRemoteMock).not.toHaveBeenCalled();
        });

        it('returns results', () => {
          expect(result).toEqual(mlClassifyLocalMockReturn);
        });
      });

      describe('when tensorflow is defined but not present', () => {
        let result;

        beforeEach(async () => {
          window.tf = {};
          window.use = {
            load: () => {},
          };

          result = await mlClassify(config, input);
        });

        it('calls remote classify', () => {
          expect(mlClassifyRemoteMock).toHaveBeenCalledWith(config, input);
        });

        it('does not call local classify', () => {
          expect(mlClassifyLocalMock).not.toHaveBeenCalled();
        });

        it('returns results', () => {
          expect(result).toEqual(mlClassifyRemoteMockReturn);
        });
      });

      describe('when tensorflow is not present', () => {
        let result;

        beforeEach(async () => {
          window.use = {
            load: () => {},
          };

          result = await mlClassify(config, input);
        });

        it('calls remote classify', () => {
          expect(mlClassifyRemoteMock).toHaveBeenCalledWith(config, input);
        });

        it('does not call local classify', () => {
          expect(mlClassifyLocalMock).not.toHaveBeenCalled();
        });

        it('returns results', () => {
          expect(result).toEqual(mlClassifyRemoteMockReturn);
        });
      });
    });

    describe('when USE is defined but not present', () => {
      let result;

      beforeEach(async () => {
        window.tf = {
          loadLayersModel: () => {},
        };
        window.use = {};

        result = await mlClassify(config, input);
      });

      it('calls remote classify', () => {
        expect(mlClassifyRemoteMock).toHaveBeenCalledWith(config, input);
      });

      it('does not call local classify', () => {
        expect(mlClassifyLocalMock).not.toHaveBeenCalled();
      });

      it('returns results', () => {
        expect(result).toEqual(mlClassifyRemoteMockReturn);
      });
    });

    describe('when USE is not present', () => {
      let result;

      beforeEach(async () => {
        window.tf = {
          loadLayersModel: () => {},
        };

        result = await mlClassify(config, input);
      });

      it('calls remote classify', () => {
        expect(mlClassifyRemoteMock).toHaveBeenCalledWith(config, input);
      });

      it('does not call local classify', () => {
        expect(mlClassifyLocalMock).not.toHaveBeenCalled();
      });

      it('returns results', () => {
        expect(result).toEqual(mlClassifyRemoteMockReturn);
      });
    });
  });

  describe('mlClassifyRemote', () => {
    const input = [
      'line 1',
      'line 2',
    ];
    const jsonReturn = [
      [0, 0, 1],
      [1, 0, 0],
    ];
    let fetchMock;
    const fetchMockReturn = {
      json: () => Promise.resolve(jsonReturn),
    };
    let result;

    beforeEach(() => {
      fetchMock = jest.fn().mockResolvedValue(fetchMockReturn);
      window.fetch = fetchMock;
    });

    afterEach(() => {
      config.options.mlClassifyEndpoint = undefined;
      window.fetch = undefined;
    });

    describe('when mlClassifyEndpoint is set', () => {
      beforeEach(async () => {
        config.options.mlClassifyEndpoint = 'https://example.com/test';

        result = await mlClassifyRemote(config, input);
      });

      it('sends fetch request', () => {
        expect(fetchMock.mock.calls).toMatchSnapshot();
      });

      it('returns response from fetch', () => {
        expect(result).toEqual(jsonReturn);
      });
    });

    describe('when mlClassifyEndpoint is not set', () => {
      it('should throw', async () => {
        expect(mlClassifyRemote(config, input)).rejects.toThrow();
      });
    });
  });

  describe('mlClassifyLocal', () => {
    const input = [
      'line 1',
      'line 2',
    ];
    const mockEmbeddedSentence = [131, 1551, 156]; // Some vector
    const predictionValue = [0, 0, 1];
    // TF
    let loadModelMock;
    let predictMock;
    let dataSyncMock;

    // USE
    let loadUseMock;
    let embedMock;

    let result;

    beforeEach(async () => {
      dataSyncMock = jest.fn().mockReturnValue(predictionValue);

      predictMock = jest.fn().mockReturnValue({
        dataSync: dataSyncMock,
      });

      loadModelMock = jest.spyOn(ml, 'loadModel').mockResolvedValue({
        predict: predictMock,
      });

      embedMock = jest.fn().mockResolvedValue(mockEmbeddedSentence);

      loadUseMock = jest.fn().mockResolvedValue({
        embed: embedMock,
      });

      window.use = {
        load: loadUseMock,
      };

      result = await mlClassifyLocal(config, input);
    });

    it('calls loadModel', () => {
      expect(loadModelMock).toHaveBeenCalled();
    });

    it('calls predict', () => {
      expect(predictMock).toHaveBeenCalledTimes(2);
      expect(predictMock).toHaveBeenCalledWith(mockEmbeddedSentence);
    });

    it('calls dataSync', () => {
      expect(dataSyncMock).toHaveBeenCalled();
    });

    it('calls load for USE', () => {
      expect(loadUseMock).toHaveBeenCalled();
    });

    it('calls embed for USE', () => {
      expect(predictMock).toHaveBeenCalledTimes(2);
      expect(embedMock).nthCalledWith(1, [input[0]]);
      expect(embedMock).nthCalledWith(2, [input[1]]);
    });

    it('returns result', () => {
      // One prediction per input line
      expect(result).toEqual([predictionValue, predictionValue]);
    });
  });

  describe('loadModel', () => {
    let loadLayersModelMock;
    const exampleModelEndpoint = 'https://example.com/test.json';

    beforeEach(() => {
      loadLayersModelMock = jest.fn();

      window.tf = {
        loadLayersModel: loadLayersModelMock,
      };
    });

    afterEach(() => {
      window.tf = undefined;
    });

    describe('when mlModelEndpoint is not set', () => {
      it('throws', () => {
        expect(loadModel(config)).rejects.toThrow();
      });

      it('does not call loadLayersModel', async () => {
        try {
          await loadModel(config);
        } catch (e) {
          // Do nothing
        }

        expect(loadLayersModelMock).not.toHaveBeenCalled();
      });
    });

    describe('when mlModelEndpoint is set', () => {
      beforeEach(() => {
        config.options.mlModelEndpoint = exampleModelEndpoint;
      });

      it('calls loadLayersModel', async () => {
        await loadModel(config);

        expect(loadLayersModelMock).toHaveBeenCalledWith(exampleModelEndpoint);
      });
    });
  });

  describe('findPotentialSetsByHeader', () => {
    let getDocumentContentMock;
    const exampleDocumentContent = [
      'example 1',
      'example 2',
      'example 3',
      'example 4',
      'example 5',
    ];

    beforeEach(() => {
      getDocumentContentMock = jest.spyOn(ml, 'getDocumentContent').mockReturnValue(exampleDocumentContent);
    });

    it('calls getDocumentContent', () => {
      findPotentialSetsByHeader(config, /example 1/);
      expect(getDocumentContentMock).toHaveBeenCalled();
    });

    it('returns content found after regex', () => {
      const result = findPotentialSetsByHeader(config, /example 2/);
      expect(result).toEqual([
        [
          'example 3',
          'example 4',
          'example 5',
        ],
      ]);
    });

    it('returns multiple when regex matches', () => {
      const result = findPotentialSetsByHeader(config, /example 2|example 4/);
      expect(result).toEqual([
        [
          'example 3',
          'example 4',
          'example 5',
        ],
        [
          'example 5',
        ],
      ]);
    });
  });

  describe('getDocumentContent', () => {
    let applyLIBlockStylingMock;
    const documentHtml = '<div>example 1\nexample 2</div>';
    const documentContent = [
      'example 1',
      'example 2',
    ];
    let result;

    beforeEach(() => {
      applyLIBlockStylingMock = jest.spyOn(element, 'applyLIBlockStyling');
      document.body.innerHTML = documentHtml;

      result = getDocumentContent(config);
    });

    it('returns document innerText', () => {
      expect(result).toEqual(documentContent);
    });

    it('calls applyLIBlockStyling', () => {
      expect(applyLIBlockStylingMock).toHaveBeenCalledWith(window.document.body);
    });
  });
});
