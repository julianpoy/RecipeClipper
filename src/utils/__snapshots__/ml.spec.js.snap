// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ml mlClassifyRemote when RC_ML_CLASSIFY_ENDPOINT is set sends fetch request 1`] = `
Array [
  Array [
    "https://example.com/test",
    Object {
      "body": "{\\"sentences\\":[\\"line 1\\",\\"line 2\\"]}",
      "headers": Object {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;
